apps/api/src/migration/1676405526677-CreateReportView.ts

```diff
@@ -0,0 +1,276 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class CreateReportView1676405526677 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+      CREATE OR REPLACE VIEW hired_withdrawn_applicants AS (
+        SELECT
+          iasa.applicant_id AS id, 
+          iasa.start_date AS hired_date,
+          CAST(NULL AS date) AS withdraw_date
+        FROM public.ien_applicant_status_audit iasa
+        LEFT JOIN public.ien_applicant_status ias ON ias.id = iasa.status_id
+        WHERE ias.status = 'Job Offer Accepted'
+        UNION ALL
+        SELECT
+          iasa.applicant_id AS id, 
+          CAST(NULL AS date) AS hired_date,
+          iasa.start_date AS withdraw_date
+        FROM public.ien_applicant_status_audit iasa
+        LEFT JOIN public.ien_applicant_status ias ON ias.id = iasa.status_id
+        WHERE ias.status = 'Withdrew from IEN program'
+      );
+    `);
+    await queryRunner.query(`
+      CREATE OR REPLACE VIEW hired_withdrawn_applicant_milestone AS (
+        SELECT
+          hwa.*,
+          (SELECT min(start_date) FROM public.ien_applicant_status_audit asa WHERE asa.applicant_id = hwa.id) as milestone_start_date,
+          (SELECT min(start_date) FROM public.ien_applicant_status_audit asa
+            WHERE asa.applicant_id = hwa.id AND asa.status_id IN (
+              SELECT ias.id FROM ien_applicant_status ias WHERE ias.status IN (
+                'Applied to NNAS',
+                'Submitted Documents (NNAS Application in Review)',
+                'Received NNAS Report'
```

Comment:
would we have to re-write this migration again if they were to change the wording in some of the statuses in ATS?

Reply:
I think so. The latest `typeorm` supports `ViewEntity` in which we can define a view query. Unfortunately, it didn't work properly for the version we are using. I'll try it again.

Reply:
Would it be possible to use the id of the Milestone instead of the name? That would allow us to change the name of the milestone in future without modifying the reports.

Reply:
@dbayly-freshworks 
If we use ID instead of status, it could only work for just one database. I think we need to add a new field like 'abbreviation' or 'code' to differentiate it from status as 'display text'. I am not sure if we should do it in this ticket.

Reply:
I got ViewEntity working. I'll remove the migration.

Reply:
The Ids for each are set both in the migration, and by the ATS sync. I think we have have some level of confidence that the ids will be the same across environments. I have a JSON file that has all the ids if thats useful.

Reply:
ViewEntity doesn't generate a temporary table with `expression` cause we don't sync schema. so, we need the migration to create view tables. I defined a constant map of status and its id and made the long query more readable.